/actions/setup-python can be flakey with multiple jobs. Can pyenv versions be cached?

Hi team.  I've been using the setup-Python cation for a few years and overall its worked great.
You can have too much of a good thing though I suppose, and I've encountered the occasional sporadic failure when using the action in a large test matrix, 10 - 20 jobs overall, when pyenv hits a rate limit, in particular from https://api.github.com/repos/oracle/graalpython/releases.  

These failures aren't LizardByte's fault - Github generously provide lots of test runners.  But when these runners en masse retrieve test data artefacts etc., they suffer from rate limits and anti bot measures, even though they're trying to download from their own server.  

But I wondered if it was possible to cache the Python releases elsewhere?  I suspect the vast majority of your other users have no need to test on GraalPython either, so if an option could be added to tell Pyenv to ignore GraalPy, that might work too.

Many thanks,

James


Example failure to to Github's flakey downloads:

```
Mode                 LastWriteTime         Length Name                                                                 
----                 -------------         ------ ----                                                                 
d-----          6/7/2026   2:39 PM                .pyenv                                                               
pyenv-win is successfully installed. You may need to close and reopen your terminal before using it.


Run pyenv update
:: [Info] ::  Mirror: https://www.python.org/ftp/python
:: [Info] ::  Mirror: https://downloads.python.org/pypy/versions.json
:: [Info] ::  Mirror: https://api.github.com/repos/oracle/graalpython/releases
HTTP Error downloading from mirror "https://api.github.com/repos/oracle/graalpython/releases"
Error(403): rate limit exceeded
Error: Process completed with exit code 1.


```

<img width="990" height="750" alt="Image" src="https://github.com/user-attachments/assets/d8d20742-a153-4887-8025-3e1ca8a8fcd7" />

--- a/actions/setup_python/README.md
+++ b/actions/setup_python/README.md
@@ -64,6 +64,7 @@ steps:
 | python-version      | The version(s) of Python to set up. Can be a single version or multiple versions separated by newlines or spaces.        |         | `false`  |
 | python-version-file | File containing the Python version to set up. Supports `.python-version`, `pyproject.toml`, `.tool-versions`, `Pipfile`. |         | `false`  |
 | update-environment  | Whether to update the environment variables.                                                                             | `true`  | `false`  |
+| update-pyenv  | Whether Pyenv should retrieve the list of latest available Pythons variables.                                                                             | `true`  | `false`  |
 
 > [!NOTE]
 > Either `python-version` or `python-version-file` must be specified.

--- a/actions/setup_python/README.md
+++ b/actions/setup_python/README.md
@@ -64,7 +64,7 @@ steps:
 | python-version      | The version(s) of Python to set up. Can be a single version or multiple versions separated by newlines or spaces.        |         | `false`  |
 | python-version-file | File containing the Python version to set up. Supports `.python-version`, `pyproject.toml`, `.tool-versions`, `Pipfile`. |         | `false`  |
 | update-environment  | Whether to update the environment variables.                                                                             | `true`  | `false`  |
-| update-pyenv  | Whether Pyenv should retrieve the list of latest available Pythons variables.                                                                             | `true`  | `false`  |
+| update-pyenv        | Whether Pyenv should retrieve the list of latest available Pythons variables.                                            | `true`  | `false`  |
 
 > [!NOTE]
 > Either `python-version` or `python-version-file` must be specified.

--- a/actions/setup_python/action.yml
+++ b/actions/setup_python/action.yml
@@ -22,6 +22,10 @@ inputs:
     description: Whether to update the environment variables.
     default: "true"
     required: false
+  update-pyenv:
+    description: Whether to update available Python versions (set to false to avoid flakey errors caused by third party rate limits).
+    default: "true"
+    required: false
 
 outputs:
   python-version:
@@ -67,6 +71,7 @@ runs:
         done
 
     - name: Update pyenv
+      if: ${{ inputs.update-pyenv == 'true' }}
       shell: bash
       run: pyenv update
 

--- a/actions/setup_python/action.yml
+++ b/actions/setup_python/action.yml
@@ -23,7 +23,9 @@ inputs:
     default: "true"
     required: false
   update-pyenv:
-    description: Whether to update available Python versions (set to false to avoid errors from third party rate limits).
+    description: >
+      Whether to update available Python versions.
+      Set to false to avoid errors from third party rate limits.
     default: "true"
     required: false
 

--- a/actions/setup_python/action.yml
+++ b/actions/setup_python/action.yml
@@ -23,7 +23,7 @@ inputs:
     default: "true"
     required: false
   update-pyenv:
-    description: Whether to update available Python versions (set to false to avoid flakey errors caused by third party rate limits).
+    description: Whether to update available Python versions (set to false to avoid errors from third party rate limits).
     default: "true"
     required: false
 

